debian/patches/send-mpls-routes-from-zebra-FPM.patch

```diff
@@ -0,0 +1,1320 @@
+-		 * When a user adds a new protocol this will show up
+-		 * to let them know to do something about it.  This
+-		 * is intentionally a warn because we should see
+-		 * this as part of development of a new protocol
+-		 */
+-		zlog_debug(
+-			"%s: Please add this protocol(%d) to proper rt_netlink.c handling",
+-			__PRETTY_FUNCTION__, proto);
+-		proto = RTPROT_ZEBRA;
+-		break;
+-	}
+-
+-	return proto;
+-}
+ 
+ static inline int proto2zebra(int proto, int family, bool is_nexthop)
+ {
+diff --git a/zebra/rt_netlink.h b/zebra/rt_netlink.h
+index 2b4b14514..3f0d81e37 100644
+--- a/zebra/rt_netlink.h
++++ b/zebra/rt_netlink.h
+@@ -61,6 +61,76 @@ extern "C" {
+ #define RTPROT_ZSTATIC     196
+ #define RTPROT_OPENFABRIC  197
+ 
++struct gw_family_t {
++	uint16_t filler;
++	uint16_t family;
++	union g_addr gate;
++};
++
++static inline int zebra2proto(int proto)
++{
++	switch (proto) {
++	case ZEBRA_ROUTE_BABEL:
++		proto = RTPROT_BABEL;
++		break;
++	case ZEBRA_ROUTE_BGP:
++		proto = RTPROT_BGP;
++		break;
++	case ZEBRA_ROUTE_OSPF:
++	case ZEBRA_ROUTE_OSPF6:
++		proto = RTPROT_OSPF;
++		break;
++	case ZEBRA_ROUTE_STATIC:
++		proto = RTPROT_ZSTATIC;
++		break;
++	case ZEBRA_ROUTE_ISIS:
++		proto = RTPROT_ISIS;
++		break;
++	case ZEBRA_ROUTE_RIP:
++		proto = RTPROT_RIP;
++		break;
++	case ZEBRA_ROUTE_RIPNG:
++		proto = RTPROT_RIPNG;
++		break;
++	case ZEBRA_ROUTE_NHRP:
++		proto = RTPROT_NHRP;
++		break;
++	case ZEBRA_ROUTE_EIGRP:
++		proto = RTPROT_EIGRP;
++		break;
++	case ZEBRA_ROUTE_LDP:
++		proto = RTPROT_LDP;
++		break;
++	case ZEBRA_ROUTE_SHARP:
++		proto = RTPROT_SHARP;
++		break;
++	case ZEBRA_ROUTE_PBR:
++		proto = RTPROT_PBR;
++		break;
++	case ZEBRA_ROUTE_OPENFABRIC:
++		proto = RTPROT_OPENFABRIC;
++		break;
++	case ZEBRA_ROUTE_TABLE:
++		proto = RTPROT_ZEBRA;
++		break;
++	default:
++		/*
++		 * When a user adds a new protocol this will show up
++		 * to let them know to do something about it.  This
++		 * is intentionally a warn because we should see
++		 * this as part of development of a new protocol
++		 */
++		zlog_debug(
++			"%s: Please add this protocol(%d) to proper rt_netlink.c handling",
++			__PRETTY_FUNCTION__, proto);
++		proto = RTPROT_ZEBRA;
++		break;
++	}
++
++	return proto;
++}
++
++
+ void rt_netlink_init(void);
+ 
+ /* MPLS label forwarding table change, using dataplane context information. */
+diff --git a/zebra/zebra_fpm.c b/zebra/zebra_fpm.c
+index fa48c03c7..adf160c32 100644
+--- a/zebra/zebra_fpm.c
++++ b/zebra/zebra_fpm.c
+@@ -34,6 +34,7 @@
+ 
+ #include "zebra/rib.h"
+ #include "zebra/zserv.h"
++#include "zebra/zebra_mpls.h"
+ #include "zebra/zebra_ns.h"
+ #include "zebra/zebra_vrf.h"
+ #include "zebra/zebra_errors.h"
+@@ -82,6 +83,17 @@ typedef struct zfpm_rnodes_iter_t_ {
+ } zfpm_rnodes_iter_t;
+ 
+ /*
++ * Structure that holds state for iterating over all lsp
++ * structures that are candidates for being communicated to the FPM.
++ */
++typedef struct zfpm_lsp_iter_t_ {
++ route_table_iter_state_t state;
```

Comment:
Please use indentation which matches the existing source files, in most cases that should be a \<tab\>. The indentation changes make it more difficult to compare this new patch to the original.